python/ray/data/tests/test_resource_manager.py

```diff
@@ -138,3 +145,145 @@ def test_calculating_usage(self):
+        resource_manager.get_global_limits = MagicMock(
+            side_effect=mock_get_global_limits
+        )
+
+        assert resource_manager.op_resource_limiter_enabled()
+        op_resource_limiter = resource_manager._op_resource_limiter
+        assert isinstance(op_resource_limiter, ReservationOpResourceLimiter)
+
+        # Test initial state when no resources are used.
+        global_limits = ExecutionResources(cpu=16, gpu=0, object_store_memory=1000)
+        op_resource_limiter.update_usages()
+        assert o1 not in op_resource_limiter._op_reserved
+        assert o4 not in op_resource_limiter._op_reserved
+        assert op_resource_limiter._op_reserved[o2] == ExecutionResources(4, 0, 250)
+        assert op_resource_limiter._op_reserved[o3] == ExecutionResources(4, 0, 250)
+        assert op_resource_limiter._total_shared == ExecutionResources(8, 0, 500)
+
+        assert op_resource_limiter.get_op_limits(o1) == ExecutionResources.inf()
+        assert op_resource_limiter.get_op_limits(o4) == ExecutionResources.inf()
+        assert op_resource_limiter.get_op_limits(o2) == ExecutionResources(
+            8, float("inf"), 500
+        )
+        assert op_resource_limiter.get_op_limits(o3) == ExecutionResources(
+            8, float("inf"), 500
+        )
+
+        # Test when each operator uses some resources.
+        op_usages[o2] = ExecutionResources(6, 0, 500)
+        op_usages[o3] = ExecutionResources(2, 0, 125)
+        op_usages[o4] = ExecutionResources(0, 0, 50)
+
+        op_resource_limiter.update_usages()
+        assert op_resource_limiter.get_op_limits(o1) == ExecutionResources.inf()
+        assert op_resource_limiter.get_op_limits(o4) == ExecutionResources.inf()
+        assert op_resource_limiter.get_op_limits(o2) == ExecutionResources(
+            3, float("inf"), 100
+        )
+        assert op_resource_limiter.get_op_limits(o3) == ExecutionResources(
+            5, float("inf"), 225
+        )
+
+        # Test global_limits updated.
+        global_limits = ExecutionResources(cpu=12, gpu=0, object_store_memory=800)
+        op_resource_limiter.update_usages()
+        assert o1 not in op_resource_limiter._op_reserved
+        assert o4 not in op_resource_limiter._op_reserved
+        assert op_resource_limiter._op_reserved[o2] == ExecutionResources(3, 0, 200)
+        assert op_resource_limiter._op_reserved[o3] == ExecutionResources(3, 0, 200)
+        assert op_resource_limiter._total_shared == ExecutionResources(6, 0, 400)
+
+        assert op_resource_limiter.get_op_limits(o1) == ExecutionResources.inf()
+        assert op_resource_limiter.get_op_limits(o4) == ExecutionResources.inf()
+        assert op_resource_limiter.get_op_limits(o2) == ExecutionResources(
+            1.5, float("inf"), 25
+        )
+        assert op_resource_limiter.get_op_limits(o3) == ExecutionResources(
+            2.5, float("inf"), 100
+        )
+
+        # Test global_limits exceeded.
```

Comment:
Nit: global_limits isn't exceeded in this test, right? 500 + 125 + 150 = 775 < 800? This is testing what happens when the shared memory is full?